### ISSUE TITLE
KeyError: 'agedge: no key'

### ISSUE THREAD

**Issue by fpettinari** [pytransitions/transitions#258]
On Windows 10 & Python 2.7
After MANY installation errors with pygraphviz I've come to this one which it seems I'm not able to solve by myself:

```
  File "C:\Python27\lib\site-packages\transitions\extensions\diagrams.py", line 332, in __init__
    model.get_graph()
  File "C:\Python27\lib\site-packages\transitions\extensions\diagrams.py", line 345, in _get_graph
    else Graph(self).get_graph(title)
  File "C:\Python27\lib\site-packages\transitions\extensions\diagrams.py", line 172, in get_graph
    self._add_edges(self.machine.events.copy(), fsm_graph)
  File "C:\Python27\lib\site-packages\transitions\extensions\diagrams.py", line 130, in _add_edges
    container.add_edge(src, dst, **edge_attr)
  File "C:\Python27\lib\site-packages\pygraphviz\agraph.py", line 478, in add_edge
    eh = gv.agedge(self.handle, uh, vh, str(key), _Action.find)
KeyError: 'agedge: no key'
```

it seems it is caused by some transitions configurations like this:

```
	['any_trigger', 'same_state', 'same_state'],
```
or
```
	['anything', '*', 'any_state'],
```

**Comment by aleneum** [comment]
Hello @fedpet,

this is a reoccurring  problem of `pygraphviz` under Windows. It has been reported before (#133), asked before at [SO](https://stackoverflow.com/questions/35610736/passing-a-string-to-agedge-in-agraph-py-issue-with-networkx-and-pygraphviz) and also reported to pygraphviz (pygraphviz/pygraphviz#88) and networkx (networkx/networkx#2012). Since the issue has been closed in the pygraphviz tracker, I assume that might be fixed `pygraphviz` version 1.4rc1.  In a pull request (#217) a user mentions a workaround for that issue.

Long story short: afaik we cannot do much about it since it seems to be related to `pygraphviz` and Windows, not `transitions`.

**Comment by fedpet** [comment]
the error only happens when I try to draw reflexive transitions or ones which have '*' as the source.
btw if the problem is pygraphviz then what version are you officially using?
I see the repo examples are working..

**Comment by aleneum** [comment]
> btw if the problem is pygraphviz then what version are you officially using?

It shouldn't matter (we have used a couple of versions) as the issue is exclusive to Windows. Our test servers run Ubuntu 14.04  and I work with OSX Sierra (both use pygraphviz 1.3.1 currently). I have also used `transitions` on Ubuntu 16.04. Never had this issue. This code:

```python
from transitions.extensions import GraphMachine

states = ['first', 'second']
transitions = [
    ['any_trigger', 'first', 'first'],
    ['anything', '*', 'second'],
]

machine = GraphMachine(states=states, transitions=transitions, initial='first',
                       auto_transitions=False, show_conditions=True)

machine.get_graph().draw('fsm.png', prog='dot')
```

produces this result:

![fsm](https://user-images.githubusercontent.com/205986/31043984-d25eb01e-a5c6-11e7-8f67-05bc1361e4b9.png)

I assume this would not work in your environment.

> pygraphviz and it should be version 1.5.dev same error.. tried with version 1.4rc1.. same error

Bummer, it seems like they closed the issue without actually fixing it then. Thanks for giving it a try.

**Comment by aleneum** [comment]
I am trying to set up a Windows CI test job but installing pygraphviz does not play very nice.

@fedpet: How did you install pygraphviz? Unfortunately, using pip after installing Graphviz finishes but I end up with DLL import errors (event though PATH is set correctly). The only way I got it to work was by using the unofficial binary builds that are suggested occasionally but this does not look like a way suitable for an automatic build process.

**Comment by aleneum** [comment]
Nevermind, PATH has not been set correctly. Our tests also fail on Windows as you can see [here](https://ci.appveyor.com/project/aleneum/transitions/build/1.0.15). I also added some try/catch phrases in the `dev-appveyor` branch to get some error messages. As mentioned in the above references issues/posts this error occurs only when source and destination of an edge are the same. I tried to omit extra attributes like labels but this does not work as well.

If you have any suggestions about how to approach this issue I am all ears.

**Comment by fedpet** [comment]
hi aleneum, thank you for your support.
Your answer suggest me the problem may be on pygraphviz or graphviz side.

A minimal code example demonstrating this bug with their libraries would be needed to be able to open a proper bug report on [pygraphviz/pygraphviz](https://github.com/pygraphviz/pygraphviz)  or [graphviz-issue-tracker](http://www.graphviz.org/content/graphviz-issue-tracker)

**Comment by aleneum** [comment]
Hi @fedpet,

I added a minimal test to `dev-appveyor`:

```python
import pygraphviz as pgv

from unittest import TestCase

class TestGraphviz(TestCase):

    def test_self_reference(self):
        G = pgv.AGraph()
        G.add_node('a')
        G.add_edge('a', 'a')
        print(G)
```

Status *nix: [works](https://travis-ci.org/pytransitions/transitions/builds/282154699)
Status Windows: [fails](https://ci.appveyor.com/project/aleneum/transitions/build/1.0.17#L445)

**Comment by fedpet** [comment]
pygraphviz/pygraphviz#137

**Comment by aleneum** [comment]
For everyone having similar issues in 2020:

I tried the minimal example ([ref](https://github.com/pytransitions/transitions/issues/258#issuecomment-333473992)) and the graph ([ref](https://github.com/pytransitions/transitions/issues/258#issuecomment-333293158)) with Graphviz 2.41 64-bit from [here](https://ci.appveyor.com/project/ellson/graphviz-pl238) (link is also mentioned on the Graphviz page [here](https://graphviz.org/download/)) and pygraphviz 1.6.dev0 with the patch mentioned [here](https://github.com/Kagami/pygraphviz/commit/fe442dc16accb629c3feaf157af75f67ccabbd6e) (see [this](https://github.com/pygraphviz/pygraphviz/issues/74#issuecomment-238323405) discussion).

Both seem to work. If you experience issues with `pygraphviz` you can use `transitions` with the `graphviz` package as mentioned in the ReadMe:

```python
from transitions.extensions import GraphMachine as Machine
m = Model()
# when you want to use graphviz explicitely
machine = Machine(model=m, use_pygraphviz=False)
```
